src/test/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsServiceTest.kt

```diff
@@ -14,25 +14,122 @@
 
 package org.wfanet.measurement.kingdom.service.api.v2alpha
 
+import com.google.common.truth.extensions.proto.ProtoTruth.assertThat
+import com.google.protobuf.ByteString
+import com.nhaarman.mockitokotlin2.UseConstructor
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.whenever
 import kotlin.test.assertFailsWith
 import kotlinx.coroutines.runBlocking
+import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.junit.runners.JUnit4
 import org.wfanet.measurement.api.v2alpha.AppendLogEntryRequest
 import org.wfanet.measurement.api.v2alpha.CreateExchangeStepAttemptRequest
+import org.wfanet.measurement.api.v2alpha.ExchangeStepAttempt
 import org.wfanet.measurement.api.v2alpha.FinishExchangeStepAttemptRequest
+import org.wfanet.measurement.common.grpc.testing.GrpcTestServerRule
+import org.wfanet.measurement.common.identity.externalIdToApiId
+import org.wfanet.measurement.common.testing.verifyProtoArgument
+import org.wfanet.measurement.internal.kingdom.AppendLogEntryRequest as InternalAppendLogEntryRequest
+import org.wfanet.measurement.internal.kingdom.ExchangeStepAttempt as InternalExchangeStepAttempt
+import org.wfanet.measurement.internal.kingdom.ExchangeStepAttemptsGrpcKt.ExchangeStepAttemptsCoroutineImplBase
+import org.wfanet.measurement.internal.kingdom.ExchangeStepAttemptsGrpcKt.ExchangeStepAttemptsCoroutineStub
+
+private const val RECURRING_EXCHANGE_ID = 1L
+private const val STEP_INDEX = 1
+private const val ATTEMPT_NUMBER = 3
+private val ARBITRARY_BYTES = ByteString.copyFromUtf8("some arbitrary bytes")
+
+private val INTERNAL_EXCHANGE_STEP_ATTEMPT: InternalExchangeStepAttempt =
+  InternalExchangeStepAttempt.newBuilder().apply {
+    externalRecurringExchangeId = RECURRING_EXCHANGE_ID
+    dateBuilder.apply {
+      year = 2021
+      month = 3
+      day = 14
+    }
+    stepIndex = STEP_INDEX
+    attemptNumber = ATTEMPT_NUMBER
+    state = InternalExchangeStepAttempt.State.ACTIVE
+    detailsBuilder.apply {
+      startTimeBuilder.seconds = 123
+      updateTimeBuilder.seconds = 456
+      addDebugLogEntriesBuilder().apply {
+        timeBuilder.seconds = 789
+        message = "some message"
+      }
+      addDebugLogEntriesBuilder().apply {
+        timeBuilder.seconds = 101112
+        message = "some other message"
+      }
+      addSharedOutputs(ARBITRARY_BYTES)
+    }
+  }.build()
+
+private val EXCHANGE_STEP_ATTEMPT: ExchangeStepAttempt = ExchangeStepAttempt.newBuilder().apply {
+  keyBuilder.apply {
+    recurringExchangeId = externalIdToApiId(RECURRING_EXCHANGE_ID)
+    exchangeId = "2021-03-14"
```

Comment:
can you add in a helper function to derive this and https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/18/files#diff-0f2817d16ddd1f92a0281232c352fcac9eb7aac7ba8e73492fa5052fa5408178R50 from common source?

Reply:
That feels a little sketchy because it's essentially reimplementing the code under test... I'd push back on this under the spirit of https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html. What do you think?

In the latest commit, I've done this with

Reply:
I don't feel strongly either way on this one. That's fine.

Reply:
I'll update this to just have the two constants spelled out explicitly, but right next to one another. This should make it easier to update both at once. (It's also not a big deal if they get out of sync because the test will immediately fail and should be fairly easy to debug.)